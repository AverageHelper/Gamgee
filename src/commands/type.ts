import type { Command } from "./Command";
import { deleteMessage } from "../actions/messages/deleteMessage";

const type: Command = {
  name: "t",
  description: "Start typing. :wink:",
  async execute({ message, client, logger }) {
    const channel = message.channel;

    logger.debug(
      `${client.user?.username.concat(" is") ?? "I am"} typing in channel ${channel.id}...`
    );
    await deleteMessage(message, "Spam: Users don't need to see this command run");

    setTimeout(() => {
      channel.stopTyping(true);
    }, 5000);

    await channel.startTyping();
    logger.debug(`Finished typing in channel ${channel.id}`);
  }
};

export default type;
