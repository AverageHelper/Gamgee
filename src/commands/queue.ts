import type { Command } from "./index";
import { SAFE_PRINT_LENGTH } from "../constants";
import { ConfigValue } from "../constants/config";
import { useQueue } from "../actions/queue/useQueue";
import { setConfigQueueChannel } from "../actions/config/setConfigValue";
import { useLogger } from "../logger";
import getQueueChannel from "../actions/queue/getQueueChannel";
import getChannelFromMention from "../helpers/getChannelFromMention";
import StringBuilder from "../helpers/StringBuilder";
import durationString from "../helpers/durationString";

const logger = useLogger();

const name = "queue";
const ARG_START = "open";
const ARG_STOP = "close";
const ARG_RESTART = "restart";
const ARG_LIMIT = "limit";

type Argument = typeof ARG_START | typeof ARG_STOP | typeof ARG_RESTART | typeof ARG_LIMIT;

const allSubargs = [ARG_START, ARG_STOP, ARG_RESTART, ARG_LIMIT];
const subargsList = allSubargs.map(v => `\`${v}\``).join(", ");

const ARG_ENTRY_DURATION = "entry-duration";
const ARG_SUB_COOLDOWN = "cooldown";

type LimitKey = typeof ARG_ENTRY_DURATION | typeof ARG_SUB_COOLDOWN;

const allLimits = [ARG_ENTRY_DURATION, ARG_SUB_COOLDOWN];
const limitsList = allLimits.map(l => `\`${l}\``).join(", ");

function isLimitKey(value: unknown): value is LimitKey {
  return !!value && typeof value === "string" && allLimits.includes(value);
}

const queue: Command = {
  name,
  description: "Manage a request queue. *(Server owner only. No touch!)*",
  uses: [
    [name, "Reports the status of the current queue."],
    [
      `${name} ${ARG_START} <channel name>`,
      "Sets the channel up as a new queue. Any existing queue is saved, but queue and request commands will go to this new queue instead."
    ],
    [`${name} ${ARG_STOP}`, "Closes the current queue."],
    [`${name} ${ARG_RESTART}`, "Empties the queue and starts a fresh queue session."],
    [`${name} ${ARG_LIMIT} [${allLimits.join("|")}]`, "Sets a limit value on the queue."]
  ],
  async execute(context) {
    const { message, args, storage } = context;

    async function reply(msg: string) {
      await message.channel.send(msg);
    }
    async function reply_private(msg: string) {
      await message.author.send(`(Reply from <#${message.channel.id}>)\n${msg}`);
    }

    // Only the guild owner may touch the queue.
    // FIXME: Add more grannular access options
    if (!message.guild?.owner?.user.tag || message.author.tag !== message.guild.owner.user.tag) {
      return reply("YOU SHALL NOT PAAAAAASS!\nOr, y'know, something like that...");
    }

    if (args.length < 1) {
      // Get the current queue's status
      const channel = await getQueueChannel(context);
      if (!channel) {
        return reply(
          `No queue is set up. Would you like to start one? (Try using \`${name} ${ARG_START}\`)`
        );
      }
      const queueIsCurrent = message.channel.id === channel.id;
      const queue = await useQueue(channel);
      const [count, playtime] = await Promise.all([queue.count(), queue.playtime()]);

      const responseBuilder = new StringBuilder();
      responseBuilder.push(`Queue channel: <#${channel.id}>`);
      if (queueIsCurrent) {
        responseBuilder.push(" (in here)");
      }
      responseBuilder.pushNewLine();

      if (count) {
        const singular = count === 1;
        const are = singular ? "is" : "are";
        const s = singular ? "" : "s";

        responseBuilder.push(`There ${are} `);
        responseBuilder.pushBold(`${count} song${s}`);
        responseBuilder.push(" in the queue, with a total playtime of ");
        responseBuilder.pushBold(`${durationString(playtime)}`);
        responseBuilder.push(".");
      } else {
        responseBuilder.push("Nothing has been added yet.");
      }
      const response = responseBuilder.result();
      await Promise.all([
        queueIsCurrent ? reply(response) : reply_private(response), //
        message.delete()
      ]);
      return;
    }

    const command = args[0] as Argument;

    switch (command) {
      case ARG_START: {
        if (args.length < 2) {
          return reply(`Please name a text channel to use for the queue!`);
        }
        const channelName = args[1];
        const channel = getChannelFromMention(message, channelName);
        if (!channel) {
          return reply(
            "That's not a real channel, or I don't know how to find it yet. Mention the channel with `#`."
          );
        }

        if (!channel.isText()) {
          return reply("I can't queue in a voice channel. Please specify a text channel instead");
        }

        logger.info(`Setting up channel '${channel.name}' for queuage.`);
        await Promise.all([
          setConfigQueueChannel(storage, channel.id),
          channel.send("This is a queue now. :smiley:"),
          message
            .delete({ reason: "Users don't need to see this command once it's run." })
            .catch(error =>
              logger.error(
                `I don't seem to have permission to delete messages: ${JSON.stringify(
                  error,
                  undefined,
                  2
                )}`
              )
            )
        ]);

        const queueIsCurrent = message.channel.id === channel?.id;
        if (!queueIsCurrent) {
          return reply(`The queue is now open! :smiley:`);
        }
        return;
      }

      case ARG_STOP: {
        const channel = await getQueueChannel(context);
        if (!channel) {
          return reply("There is no open queue to close, silly! :stuck_out_tongue:");
        }
        const queueIsCurrent = message.channel.id === channel?.id;
        const promises: Array<Promise<unknown>> = [
          setConfigQueueChannel(storage, null),
          message
            .delete({ reason: "Users don't need to see this command once it's run." })
            .catch(error =>
              logger.error(
                `I don't seem to have permission to delete messages: ${JSON.stringify(
                  error,
                  undefined,
                  2
                )}`
              )
            )
        ];
        if (channel && !queueIsCurrent) {
          promises.push(channel.send("This queue is closed. :wave:"));
        }
        await Promise.all(promises);
        return reply("The queue is now closed. :wave:");
      }

      case ARG_RESTART: {
        const channel = await getQueueChannel(context);
        if (!channel) {
          return reply("No queue is set up. Maybe that's what you wanted...?");
        }
        const queue = await useQueue(channel);
        const deleteMessages = (await queue.getAllEntries())
          .map(entry => entry.queueMessageId)
          .map(messageId => channel.messages.delete(messageId));
        await Promise.all(deleteMessages);
        await queue.clear();
        return reply("The queue has restarted.");
      }

      case ARG_LIMIT: {
        const channel = await getQueueChannel(context);
        if (!channel) {
          return reply("No queue is set up yet.");
        }
        const queue = await useQueue(channel);

        // Set limits on the queue
        if (args.length < 2) {
          return reply(`Gonna need more info than that. Add one of: ${limitsList}.`);
        }

        const limitKey = args[1];
        if (!isLimitKey(limitKey)) {
          const that = limitKey.length <= SAFE_PRINT_LENGTH ? `'${limitKey}'` : "that";
          return reply(`I'm not sure what ${that} is. ` + limitsList);
        }

        let value: ConfigValue | undefined = args.length >= 3 ? args[2] : undefined;

        switch (limitKey) {
          case ARG_ENTRY_DURATION: {
            if (value === undefined) {
              value = (await queue.getConfig()).entryDurationSeconds;
              if (value === null) {
                return reply(`There is no limit on entry duration.`);
              }
              return reply(`Entry duration limit is **${durationString(value)}**`);
            }

            // Set the guild's queue entry duration limit
            value = args[2] === "null" ? null : parseInt(args[2]);
            if (value !== null && isNaN(value)) {
              value = (await queue.getConfig()).entryDurationSeconds;
              await reply("That doesn't look like an integer. Enter a number value in seconds");
            }
            await queue.updateConfig({ entryDurationSeconds: value });
            if (value === null) {
              return reply(`Entry duration limit **removed**`);
            }
            return reply(`Entry duration limit set to **${durationString(value)}**`);
          }

          case ARG_SUB_COOLDOWN: {
            if (value === undefined) {
              value = (await queue.getConfig()).cooldownSeconds;
              if (value === null) {
                return reply(`There is no submission cooldown time`);
              }
              return reply(`Entry duration limit is **${durationString(value)}**`);
            }
            return reply(`Entry duration limit set to **${durationString(0)}**`);
          }

          default: {
            const that =
              (limitKey as string).length <= SAFE_PRINT_LENGTH ? `'${limitKey as string}'` : "that";
            return reply(`I'm not sure what ${that} is. ` + limitsList);
          }
        }
      }

      default:
        return reply(`Invalid command argument. Expected ${subargsList}`);
    }
  }
};

export default queue;
