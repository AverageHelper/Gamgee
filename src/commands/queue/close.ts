import type { GuildedSubcommand } from "../Command";
import { useGuildStorage } from "../../useGuildStorage";
import getQueueChannel from "../../actions/queue/getQueueChannel";

const close: GuildedSubcommand = {
	name: "close",
	description: "Stop accepting song requests to the queue.",
	type: "SUB_COMMAND",
	requiresGuild: true,
	permissions: ["owner", "admin", "queue-admin"],
	async execute({ guild, channel, type, reply, followUp, deleteInvocation }) {
		const guildStorage = useGuildStorage(guild);
		const [isQueueOpen, queueChannel] = await Promise.all([
			guildStorage.isQueueOpen(),
			getQueueChannel(guild),
			deleteInvocation()
		]);

		if (!queueChannel) {
			return reply({
				content: "There's no queue to close. Have you set one up yet?",
				ephemeral: true
			});
		}
		if (!isQueueOpen) {
			return reply({
				content: "The queue is already closed, silly! :stuck_out_tongue:",
				ephemeral: true
			});
		}

		const queueIsCurrent = channel?.id === queueChannel.id;
		const promises: Array<Promise<unknown>> = [guildStorage.setQueueOpen(false)];
		if (!queueIsCurrent) {
			promises.push(queueChannel.send("This queue is closed. :wave:"));
		}
		if (type === "interaction") {
			promises.push(reply({ content: "Got it!", ephemeral: true }));
		}
		await Promise.all(promises);
		await followUp({ content: "The queue is now closed. :wave:", reply: false });
	}
};

export default close;
