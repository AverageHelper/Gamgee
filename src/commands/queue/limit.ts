import type { CommandInteractionOption } from "discord.js";
import type { Subcommand } from "../Command.js";
import { SAFE_PRINT_LENGTH } from "../../constants/output.js";
import { useQueue } from "../../actions/queue/useQueue.js";
import durationString from "../../helpers/durationString.js";
import getQueueChannel from "../../actions/queue/getQueueChannel.js";
import { composed, createPartialString, push, pushBold } from "../../helpers/composeStrings.js";
import { getQueueConfig } from "../../useQueueStorage.js";
import {
	resolveIntegerFromOption,
	resolveStringFromOption
} from "../../helpers/optionResolvers.js";

type LimitKey = "entry-duration" | "cooldown" | "count";

export interface QueueLimitArg {
	name: string;
	value: LimitKey;
	description: string;
}

export const allLimits: Array<QueueLimitArg> = [
	{
		name: "Song Length",
		value: "entry-duration",
		description: "The maximum duration of a song submission."
	},
	{
		name: "Submission Cooldown",
		value: "cooldown",
		description:
			"The minimum amount of time that each user must wait between their own submissions."
	},
	{
		name: "Number of Submissions",
		value: "count",
		description: "The maximum number of submissions that each user may submit."
	}
];

const limitsList = allLimits.map(l => `\`${l.value}\``).join(", ");

function isLimitKey(value: unknown): value is LimitKey {
	return (
		Boolean(value) &&
		typeof value === "string" &&
		allLimits.map(l => l.value).includes(value as LimitKey)
	);
}

const limit: Subcommand = {
	name: "limit", // TODO: Alias this to "limits"
	description: "Set a limit value on the queue. (Time in seconds, where applicable)",
	options: [
		{
			name: "key",
			description: "The name of the limit.",
			type: "STRING",
			required: true,
			choices: allLimits
		},
		{
			name: "value",
			description: "The new value to set for the key.",
			type: "INTEGER"
		}
	],
	type: "SUB_COMMAND",
	requiresGuild: true,
	permissions: ["owner", "admin", "queue-admin"],
	async execute(context) {
		const { guild, options, reply } = context;
		const queueChannel = await getQueueChannel(guild);

		if (!queueChannel) {
			return reply("No queue is set up.");
		}

		const queue = useQueue(queueChannel);
		const config = await getQueueConfig(queueChannel);

		const keyOption: CommandInteractionOption | undefined = options.data[0];
		const valueOption: CommandInteractionOption | undefined = options.data[1];

		if (!keyOption) {
			const { default: limits } = await import("../limits");
			return limits.execute(context);
		}

		const key: string = resolveStringFromOption(keyOption);

		if (!isLimitKey(key)) {
			const that = key.length <= SAFE_PRINT_LENGTH ? `'${key}'` : "that";
			return reply(`I'm not sure what ${that} is. Try one of ${limitsList}`);
		}

		// Set limits on the queue
		switch (key) {
			case "entry-duration": {
				// Limit each duration entry
				if (!valueOption) {
					// Read the current limit
					const value = config.entryDurationSeconds;
					if (value === null) {
						return reply(`There is no limit on entry duration.`);
					}
					return reply(`Entry duration limit is **${durationString(value)}**`);
				}

				// Set a new limit
				let value = resolveIntegerFromOption(valueOption);
				if (value !== null && Number.isNaN(value)) {
					return reply("That doesn't look like an integer. Enter a number value in seconds.");
				}
				value = value === null || value <= 0 ? null : value;
				await queue.updateConfig({ entryDurationSeconds: value });

				const response = createPartialString("Entry duration limit ");
				if (value === null || value <= 0) {
					pushBold("removed", response);
				} else {
					push("set to ", response);
					pushBold(durationString(value), response);
				}
				return reply(composed(response));
			}

			case "cooldown": {
				// Limit submission cooldown
				if (!valueOption) {
					// Read the current limit
					const value = config.cooldownSeconds;
					if (value === null) {
						return reply("There is no submission cooldown time");
					}
					return reply(`Submission cooldown is **${durationString(value)}**`);
				}

				// Set a new limit
				let value = resolveIntegerFromOption(valueOption);
				if (value !== null && Number.isNaN(value)) {
					value = config.cooldownSeconds;
					return reply("That doesn't look like an integer. Enter a number value in seconds.");
				}
				value = value === null || value <= 0 ? null : value;
				await queue.updateConfig({ cooldownSeconds: value });

				const response = createPartialString("Submission cooldown ");
				if (value === null || value <= 0) {
					pushBold("removed", response);
				} else {
					push("set to ", response);
					pushBold(durationString(value), response);
				}
				return reply(composed(response));
			}

			case "count": {
				// Limit submission count per user
				if (!valueOption) {
					// Read the current limit
					const value = config.submissionMaxQuantity;
					if (value === null) {
						return reply("There is no limit on the number of submissions per user.");
					}
					return reply(`Max submissions per user is **${value}**`);
				}

				// Set a new limit
				let value = resolveIntegerFromOption(valueOption);
				if (value !== null && Number.isNaN(value)) {
					value = config.submissionMaxQuantity;
					return reply("That doesn't look like an integer. Enter a number value in seconds.");
				}
				value = value === null || value <= 0 ? null : value;
				await queue.updateConfig({ submissionMaxQuantity: value });

				const response = createPartialString("Submission count limit per user ");
				if (value === null || value <= 0) {
					pushBold("removed", response);
				} else {
					push("set to ", response);
					pushBold(`${value}`, response);
				}
				return reply(composed(response));
			}
		}
	}
};

export default limit;
