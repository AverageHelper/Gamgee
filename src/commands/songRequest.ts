import { useQueue, UnsentQueueEntry } from "../actions/useQueue";
import type { Command } from "./index";
import getVideoDetails from "../actions/getVideoDetails";
import getQueueChannel from "../actions/getQueueChannel";
import { useLogger } from "../logger";

const logger = useLogger();

const name = "sr";

const yt: Command = {
  name,
  description: "Submit a song to the queue.",
  uses: [
    [`${name} <song name or YouTube link>`, "Attempts to add the given content to the queue."]
  ],
  async execute(context) {
    const { message, args } = context;

    async function reject_public(reason: string) {
      await message.channel.send(reason);
    }

    const queueChannel = await getQueueChannel(context);
    if (!queueChannel) {
      return reject_public(
        "No queue channel has been set up yet. Ask an administrator to set one up."
      );
    }

    if (message.channel.id === queueChannel.id) {
      // TODO: The ability to configure a specific channel from which song requests should be taken.
      return reject_public("Requesting songs in the queue channel hasn't been implemented yet.");
    }

    if (args.length < 1) {
      return reject_public("You're gonna have to add a song link or title to that.");
    }

    async function accept(entry: UnsentQueueEntry, sendUrl = false) {
      if (!queueChannel) {
        return reject_public(
          "No queue channel has been set up yet. Ask an administrator to set one up."
        );
      }

      logger.debug(`Preparing queue cache for channel ${queueChannel.id} (#${queueChannel.name})`);
      const queue = await useQueue(queueChannel);
      logger.debug("Queue prepared!");
      await Promise.all([
        queue.push(entry), //
        sendUrl ? message.channel.send(entry.url) : null
      ]);
      logger.debug(
        `Pushed new entry to queue. Sending public acceptance to user ${message.author.id} (${message.author.username})`
      );
      // Send acceptance after the potential `send(entry.url)` call
      await message.channel.send(`**${message.author.username}**, Submission Accepted!`);
      logger.debug("Responded.");
    }

    try {
      const video = await getVideoDetails(args);
      if (video === null) {
        return reject_public("No songs were found!");
      }

      const url = video.url;
      const minutes = video.duration.seconds / 60;
      const sentAt = message.createdAt;
      const senderId = message.author.id;

      // Whether this is a search result and we therefore haven't had this link embedded yet
      const shouldSendUrl = "type" in video && video.type === "video";

      // Full send!
      return accept({ url, minutes, sentAt, senderId }, shouldSendUrl);

      // Handle fetch errors
    } catch (error) {
      logger.error("Failed to run query", args, error);
      return reject_public("That video query gave me an error.");
    }
  }
};

export default yt;
