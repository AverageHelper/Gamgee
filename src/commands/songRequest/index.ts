import type Discord from "discord.js";
import type { Command } from "./../index";
import { deleteMessage, replyPrivately } from "../../actions/messages";

import arbitrarySubcommand from "./urlRequest";
import info from "./info";
import setup from "./setup";
import open from "./open";
import close from "./close";
import limit from "./limit";
import stats from "./stats";
import restart from "./restart";

const namedSubcommands = [info, setup, open, close, limit, stats, restart];

export async function reply(message: Discord.Message, msg: string): Promise<void> {
  await Promise.all([
    message.channel.send(msg), //
    message.channel.stopTyping(true)
  ]);
}

export async function reply_private(message: Discord.Message, msg: string): Promise<void> {
  await replyPrivately(message, msg);
}

export async function reject_private(message: Discord.Message, reason: string): Promise<void> {
  await Promise.all([
    deleteMessage(message, "Spam; this song request was rejected."),
    replyPrivately(message, reason)
  ]);
}

export async function reject_public(message: Discord.Message, reason: string): Promise<void> {
  await Promise.all([
    message.channel.send(`:hammer: <@!${message.author.id}>, ${reason}`),
    message.suppressEmbeds(true)
  ]);
}

const sr: Command = {
  name: "sr",
  description: "Submit a song to the queue.",
  namedSubcommands,
  arbitrarySubcommand,
  async execute(context) {
    const { message, args } = context;

    // Prepare arguments
    if (args.length < 1) {
      return reject_public(message, "You're gonna have to add a song link to that.");
    }

    const arg = args[0];

    for (const command of namedSubcommands) {
      if (command.name === arg) {
        return command.execute(context);
      }
    }

    return arbitrarySubcommand.execute(context);
  }
};

export default sr;
