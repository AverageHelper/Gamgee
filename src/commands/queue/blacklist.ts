import type { Subcommand } from "../Command.js";
import { blacklistUser, getQueueConfig } from "../../useQueueStorage.js";
import { getConfigCommandPrefix } from "../../actions/config/getConfigValue.js";
import { getQueueChannel } from "../../actions/queue/getQueueChannel.js";
import { logUser } from "../../helpers/logUser.js";
import { sr as parentCommand } from "../songRequest.js";
import { resolveUserFromOption } from "../../helpers/optionResolvers.js";
import { whitelist } from "./whitelist.js";
import {
	composed,
	createPartialString,
	push,
	pushCode,
	pushNewLine
} from "../../helpers/composeStrings.js";

export const blacklist: Subcommand = {
	name: "blacklist",
	description: "Show the list of blacklisted users, or add a user to the blacklist.",
	options: [
		{
			name: "user",
			description: "Block the user from making song requests",
			type: "USER",
			required: false
		}
	],
	type: "SUB_COMMAND",
	requiresGuild: true,
	async execute(context) {
		const {
			guild,
			user,
			options,
			storage,
			logger,
			reply,
			replyPrivately,
			deleteInvocation
		} = context;

		const queueChannel = await getQueueChannel(guild);
		if (!queueChannel) {
			return await reply({ content: ":x: There's no queue set up yet.", ephemeral: true });
		}

		const firstOption = options[0];
		if (!firstOption) {
			if (context.type === "message") {
				logger.debug("Private-ness for message commands is restricted to DMs.");
				// We can reply to text messages twice, since the second one will be a DM. We `await` here, not `return`.
				await reply(":paperclip: Check the list in your DMs");
			}

			const queueConfig = await getQueueConfig(queueChannel);
			const blacklistedUsers = queueConfig.blacklistedUsers?.map(user => user.id) ?? [];

			const prefix = await getConfigCommandPrefix(storage);
			const guildName = guild.name.trim();

			const replyMsg = createPartialString();

			push(`**Song Request Blacklist for *${guildName}***`, replyMsg);
			pushNewLine(replyMsg);
			if (blacklistedUsers.length === 0) {
				push(" - Nobody. (Let's hope it remains this way.)", replyMsg);
				pushNewLine(replyMsg);
			}

			blacklistedUsers.forEach(userId => {
				push(` - <@${userId}>`, replyMsg);
				pushNewLine(replyMsg);
			});

			pushNewLine(replyMsg);
			push("To add a user to the blacklist, run ", replyMsg);
			pushCode(`${prefix}${parentCommand.name} ${blacklist.name} <user mention>`, replyMsg);
			push(".", replyMsg);

			pushNewLine(replyMsg);
			push("To remove a user from the blacklist, run ", replyMsg);
			pushCode(`${prefix}${parentCommand.name} ${whitelist.name} <user mention>`, replyMsg);
			push(".", replyMsg);
			pushNewLine(replyMsg);
			if (context.type === "message") {
				// These are DMs. Be clear about where to run commands.
				push("(Run these in ", replyMsg);
				push(`*${guildName}*`, replyMsg);
				push(", obviously)", replyMsg);
			}

			return await replyPrivately(composed(replyMsg));
		}

		await deleteInvocation();

		const subject = await resolveUserFromOption(firstOption, guild);
		if (!subject) {
			return await reply({ content: ":x: I don't know who that is.", ephemeral: true });
		}

		if (subject.id === user.id) {
			return await reply({ content: ":x: You can't blacklist yourself, silly!", ephemeral: true });
		}

		if (subject.id === guild.ownerId) {
			return await reply({
				content: ":x: I can't blacklist the owner. That would be rude!",
				ephemeral: true
			});
		}

		await blacklistUser(subject.id, queueChannel);
		logger.info(`Removed song request permission from user ${logUser(subject)}.`);

		return await reply({
			content: `:pirate_flag: <@!${subject.id}> is no longer allowed to submit song requests.`,
			shouldMention: false,
			ephemeral: true
		});
	}
};
