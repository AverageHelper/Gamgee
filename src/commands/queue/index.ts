import type { Command } from "../Command";
import StringBuilder from "../../helpers/StringBuilder";
import setup from "./setup";
import teardown from "./teardown";
import blacklist from "./blacklist";
import whitelist from "./whitelist";
import open from "./open";
import close from "./close";
import limit from "./limit";
import stats from "./stats";
import restart from "./restart";

const namedSubcommands = [
  setup,
  teardown,
  blacklist,
  whitelist,
  open,
  close,
  limit,
  stats,
  restart
];

const sr: Command = {
  name: "queue",
  description: "Submit a song to the queue.",
  options: namedSubcommands,
  async execute(context) {
    const arg: string | undefined = context.options[0]?.name;
    const argOptions = context.options[0]?.options;
    if (arg === undefined || arg === "" || !argOptions) {
      const response = new StringBuilder("The possible subcommands are:");
      Object.keys(namedSubcommands).forEach(commandName => {
        response.pushNewLine();
        response.push(" - ");
        response.pushCode(commandName);
      });

      return context.reply(response.result());
    }

    for (const command of namedSubcommands) {
      if (command.name === arg) {
        context.options = argOptions;
        return command.execute(context);
      }
    }

    // Unknown command. Leave it be
  }
};

export default sr;
