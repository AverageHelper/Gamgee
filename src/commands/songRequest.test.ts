jest.mock("../useGuildStorage");
jest.mock("../actions/queue/getQueueChannel");
jest.mock("../actions/queue/useQueue");
jest.mock("../actions/getVideoDetails");

import * as queueActions from "../actions/queue/useQueue";
const mockUseQueue = queueActions.useQueue as jest.Mock;

import * as guildStorage from "../useGuildStorage";
const mockGuildStorage = guildStorage.useGuildStorage as jest.Mock;

import getQueueChannel from "../actions/queue/getQueueChannel";
const mockGetQueueChannel = getQueueChannel as jest.Mock;

import getVideoDetails from "../actions/getVideoDetails";
const mockGetVideoDetails = getVideoDetails as jest.Mock;
mockGetVideoDetails.mockImplementation(async (url: string) => {
	// Enough uncertainty that *something* should go out of order if it's going to
	const ms = Math.floor(Math.random() * 50);
	await new Promise(resolve => setTimeout(resolve, ms));
	return {
		url,
		title: "video-title",
		duration: {
			seconds: 500
		},
		fromUrl: true
	};
});

import type { GuildedCommandContext } from "./Command";
import { useTestLogger } from "../../tests/testUtils/logger";
import Discord from "discord.js";
import songRequest from "./songRequest";

const logger = useTestLogger("error");

describe("Song request via URL", () => {
	const urls: [string, string, string, string, string, string, string, string, string, string] = [
		"https://youtu.be/dQw4w9WgXcQ",
		"https://youtu.be/9RAQsdTQIcs",
		"https://youtu.be/tao1Ic8qVkM",
		"https://youtu.be/sSukg-tAK1k",
		"https://youtu.be/9eWHXhLu-uM",
		"https://youtu.be/jeKH5HhmNQc",
		"https://youtu.be/NUYvbT6vTPs",
		"https://youtu.be/aekVhtK9yuQ",
		"https://youtu.be/BwyY5LdpECA",
		"https://youtu.be/7btMEX3kAPo"
	];
	const botId = "this-user";

	const mockPrepareForLongRunningTasks = jest.fn().mockResolvedValue(undefined);
	const mockReply = jest.fn().mockResolvedValue(undefined);
	const mockReplyPrivately = jest.fn().mockResolvedValue(undefined);
	const mockChannelSend = jest.fn().mockResolvedValue(undefined);
	const mockDeleteMessage = jest.fn().mockResolvedValue(undefined);
	const mockFollowUp = jest.fn().mockResolvedValue(undefined);

	const mockQueueGetLatestUserEntry = jest.fn().mockResolvedValue(null);
	const mockQueueUserEntryCount = jest.fn().mockResolvedValue(0);

	const mockQueuePush = jest.fn();

	mockGuildStorage.mockReturnValue({
		isQueueOpen: jest.fn().mockResolvedValue(true)
	});

	mockGetQueueChannel.mockResolvedValue({
		id: "queue-channel-123",
		name: "queue"
	});

	mockUseQueue.mockReturnValue({
		getConfig: jest.fn().mockResolvedValue({
			entryDurationSeconds: null,
			cooldownSeconds: 600,
			submissionMaxQuantity: null,
			blacklistedUsers: []
		}),
		push: mockQueuePush,
		getLatestEntryFrom: mockQueueGetLatestUserEntry,
		countFrom: mockQueueUserEntryCount
	});

	const mockClient: Discord.Client = ({ user: { id: botId } } as unknown) as Discord.Client;

	function mockMessage(senderId: string, content: string): Discord.Message {
		const mockSenderMember: Discord.GuildMember = ({
			user: { id: senderId }
		} as unknown) as Discord.GuildMember;

		return ({
			content,
			author: {
				bot: false,
				id: mockSenderMember.user.id,
				username: senderId
			},
			createdAt: new Date(),
			client: (mockClient as unknown) as Discord.Client,
			prepareForLongRunningTasks: mockPrepareForLongRunningTasks,
			reply: mockReply,
			channel: {
				id: "request-channel-456",
				send: mockChannelSend
			},
			guild: {
				members: {
					fetch: jest.fn().mockImplementation(
						(userId: string) =>
							new Promise(resolve => {
								if (userId === mockSenderMember.user.id) {
									return resolve(mockSenderMember);
								} else if (userId === botId) {
									return resolve(mockClient);
								}
							})
					)
				}
			}
		} as unknown) as Discord.Message;
	}

	describe("Song request help", () => {
		test("descibes how to submit a song", async () => {
			const context = ({
				type: "message",
				guild: "any-guild",
				channel: "any-channel",
				user: "doesn't matter",
				options: new Discord.Collection(),
				logger,
				prepareForLongRunningTasks: mockPrepareForLongRunningTasks,
				reply: mockReply,
				replyPrivately: mockReplyPrivately,
				deleteInvocation: mockDeleteMessage,
				followUp: mockFollowUp
			} as unknown) as GuildedCommandContext;

			await songRequest.execute(context);
			expect(mockReply).toHaveBeenCalledTimes(1);
			expect(mockReply).toHaveBeenCalledWith(expect.toBeString());

			const calls = mockReply.mock.calls[0] as Array<unknown>;
			const description = calls[0];
			expect(description).toMatchSnapshot();
		});
	});

	test("only a user's first submission gets in if a cooldown exists", async () => {
		const mockMessage1 = mockMessage("another-user", `?sr ${urls[0]}`);
		const mockMessage2 = mockMessage("another-user", `?sr ${urls[1]}`);

		mockQueuePush.mockImplementationOnce(() => {
			mockQueueGetLatestUserEntry.mockResolvedValueOnce({
				queueMessageId: mockMessage1.id,
				url: urls[0],
				seconds: 500,
				sentAt: new Date(),
				senderId: mockMessage1.author.id,
				isDone: false
			});
			mockQueueUserEntryCount.mockResolvedValueOnce(1);
			return Promise.resolve();
		});

		const context1 = ({
			guild: mockMessage1.guild,
			channel: mockMessage1.channel,
			user: mockMessage1.author,
			options: new Discord.Collection([
				[
					"url",
					{
						name: "url",
						value: urls[0]
					}
				]
			]),
			logger,
			prepareForLongRunningTasks: mockPrepareForLongRunningTasks,
			reply: mockReply,
			replyPrivately: mockReplyPrivately,
			deleteInvocation: mockDeleteMessage,
			followUp: mockFollowUp
		} as unknown) as GuildedCommandContext;
		const context2 = ({
			...context1,
			options: new Discord.Collection([
				[
					"url",
					{
						name: "url",
						value: urls[1]
					}
				]
			]),
			user: mockMessage2.author,
			guild: mockMessage2.guild,
			channel: mockMessage2.channel
		} as unknown) as GuildedCommandContext;

		// Request a song twice in quick succession
		void songRequest.execute(context1);
		await songRequest.execute(context2);

		// Wait for handles to close
		await new Promise(resolve => setTimeout(resolve, 500));

		// queue.push should only have been called on the first URL
		expect(mockQueuePush).toHaveBeenCalledTimes(1);
		expect(mockQueuePush).toHaveBeenCalledWith(expect.toContainEntry(["url", urls[0]]));

		// The submission should have been rejected with a cooldown warning via DMs
		expect(mockDeleteMessage).toHaveBeenCalledTimes(1);
		expect(mockReplyPrivately).toHaveBeenCalledTimes(1);
		expect(mockReplyPrivately).toHaveBeenCalledWith(expect.stringContaining("must wait"));
	});

	test("submissions enter the queue in order", async () => {
		const mockMessages: Array<Discord.Message> = [];
		urls.forEach((url, i) => {
			const userId = `user-${i + 1}`;
			const message = mockMessage(userId, `?sr ${url}`);
			mockMessages.push(message);
		});

		await Promise.all([
			mockMessages
				.map(message => {
					return ({
						options: new Discord.Collection(
							message.content
								.split(" ")
								.slice(1)
								.map(url => ({
									name: "url",
									value: url
								}))
								.map(c => [c.name, c])
						),
						guild: message.guild,
						channel: message.channel,
						user: message.author,
						logger,
						prepareForLongRunningTasks: mockPrepareForLongRunningTasks,
						reply: mockReply,
						replyPrivately: mockReplyPrivately,
						followUp: mockFollowUp
					} as unknown) as GuildedCommandContext;
				})
				.map(songRequest.execute)
		]);

		// Wait for handles to close
		await new Promise(resolve => setTimeout(resolve, 500));

		// queue.push should have been called on each URL
		urls.forEach((url, i) => {
			expect(mockQueuePush).toHaveBeenNthCalledWith(
				i + 1,
				expect.toContainEntries([
					["url", url],
					["senderId", `user-${i + 1}`]
				])
			);
		});
		expect(mockQueuePush).toHaveBeenCalledTimes(10);
	});
});
