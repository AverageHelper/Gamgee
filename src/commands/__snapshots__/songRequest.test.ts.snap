// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Song request via URL Song request help descibes how to submit a song 1`] = `
"To submit a song, type \`?sr <link>\`.
For example: \`?sr https://youtu.be/dQw4w9WgXcQ\`
I will respond with a text verification indicating your song has joined the queue!

To get a link to the current song, type \`?now-playing\` and check your DMs."
`;
